libraries/aws_ec2_instance.rb

```diff
@@ -19,42 +19,50 @@ class AwsEc2Instance < AwsResourceBase
   "
 
   def initialize(opts = {})
-    # Call the parent class constructor
-    raise ArgumentError, 'aws_ec2_instance `instance_id` or `name` must be provided' if opts.nil?
-    opts = { instance_id: opts } if opts.is_a?(String) # this preserves the original scalar interface - note that the original implementation offered scalar 'id' or tag property :name
+    opts = { instance_id: opts } if opts.is_a?(String)
     super(opts)
-    validate_parameters(%i(instance_id name))
-    @display_name = opts[:name] || opts[:instance_id]
-    if opts[:instance_id] && !opts[:instance_id].empty? # we either use the instance_id
-      raise ArgumentError, 'aws_ec2_instance `instance_id` must be a string' if !opts[:instance_id].is_a?(String)
-      raise ArgumentError, 'aws_ec2_instance `instance_id`  must be in the format "i-" followed by 8 or 17 hexadecimal characters.' if opts[:instance_id] !~ /^i\-([0-9a-f]{8})|(^i\-[0-9a-f]{17})$/
+    validate_parameters(require_any_of: %i(instance_id name))
+
+    if opts[:instance_id] && !opts[:instance_id].empty? # Use instance_id, if provided
+      if !opts[:instance_id].is_a?(String) || opts[:instance_id] !~ /^i\-([0-9a-f]{8})|(^i\-[0-9a-f]{17})$/
+        raise ArgumentError, "#{@__resource_name__}: `instance_id` must be a string in the format of 'i-' followed by 8 or 17 hexadecimal characters."
+      end
+      @display_name = opts[:instance_id]
       instance_arguments = { instance_ids: [opts[:instance_id]] }
-    else # or use the tag naming convention
-      raise ArgumentError, 'aws_ec2_instance `name` must be provided' if opts[:name].nil? || opts[:name].empty?
-      filter = { name: 'tag:Name', values: [opts[:name]] }
-      instance_arguments = { filters: [filter] }
+    elsif opts[:name] && !opts[:name].empty? # Otherwise use name, if provided
+      raise ArgumentError, "#{@__resource_name__}: `name` must be provided" if opts[:name].nil? || opts[:name].empty?
+      @display_name = opts[:name] || opts[:instance_id]
+      instance_arguments = { filters: [{ name: 'tag:Name', values: [opts[:name]] }] }
+    else
+      raise ArgumentError, "#{@__resource_name__}: either instance_id or name must be provided"
     end
 
     catch_aws_errors do
       @resp = @aws.compute_client.describe_instances(instance_arguments)
-      @instance = @resp.reservations.first.instances.first.to_h unless @resp.reservations.first.nil? || @resp.reservations.first.instances.first.nil?
-      # TODO: review whether filtering by name is useful
-      if !@instance.nil?
-        raise Inspec::Exceptions::ResourceFailed, 'Expected only one instance to be returned when filtering by name!' if @resp.reservations.first.instances.count > 1
+
+      # rubocop:disable Style/GuardClause
```

Comment:
Would be better to avoid disabling the rule in the implementation.

Reply:
fixed